expression/util.go

```diff
@@ -172,6 +172,42 @@ func ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression) Ex
 	return expr
 }
 
+// ColumnSubstitutePartPrune same as ColumnSubstitute except that it will return if the
+// input expr is substituted, the newFunctionInternal is only called if its child is substituted
+func ColumnSubstitutePartPrune(expr Expression, schema *Schema, newExprs []Expression) (bool, Expression) {
```

Comment:
The difference is `ColumnSubstitute ` will always create new scalar function expression. This pull request only substitutes `ColumnSubstitute ` to `ColumnSubstitutePartPrune` in "partition pruning" path. Though there are many direct calls to `ColumnSubstitute`, maybe we do this in another request, so that this difference is more clear

Reply:
I think the new one is better than the old one in any case.
Would merging this two function cause too much chaos in this pr?

Reply:
@winoros there will be some changes, maybe it's more clear to put this whole refactoration into other pull requests if necessary

Reply:
I'll replace the old one in this request